Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/J2ObjCParser.kt

```diff
@@ -0,0 +1,218 @@
+        "short" ->  methodParameters.add(Parameter(name = paramName, type = parseType(parameterTypes.get(i)), nsConsumed = false))
+        else -> {
+          throw NotImplementedError("Have not implemented this type yet: " + parameterTypes.get(i).className)
+        }
+      }
+    }
+    return methodParameters
+  }
+
+  /**
+   * Parses an ASM method's return type and returns a Kotlin type
+   *
+   * @param method ASM method descriptor
+   * @return Type corresponding to method's return type
+   */
+  private fun parseMethodReturnType(methodDesc: String): Type {
+    return parseType(getReturnType(methodDesc))
+  }
+
+  /**
+   * Helper function to parse ASM types and return a Kotlin type
+   *
+   * @param type ASM type
+   * @return Kotlin type
+   */
+  private fun parseType(type: org.jetbrains.org.objectweb.asm.Type): Type {
+    when (type.className) {
+      "boolean" -> return ObjCBoolType
+      "byte" -> return IntegerType(size = 1, spelling = "byte", isSigned = true)
+      "char" -> return CharType
+      "double" -> return FloatingType(size = 8, spelling = "double")
+      "float" -> return FloatingType(size = 4, spelling = "float")
+      "int" -> return IntegerType(size = 4, spelling = "int", isSigned = true)
+      "long" -> return IntegerType(size = 8, spelling = "long", isSigned = true)
+      "short" -> return IntegerType(size = 2, spelling = "short", isSigned = true)
+      "void" -> return VoidType
+      else -> {
+        throw NotImplementedError("Have not implemented this type yet: " + type.className)
+      }
+    }
+  }
+
+}
+
+/**
+ * Visits methods from J2ObjCParser and collects parameter names
+ *
+ * @param paramNames List of parameter name strings to be added to
+ */
+private class MethodBuilder(val paramNames: MutableCollection<List<String>>): MethodVisitor(Opcodes.ASM5) {
+  val params = mutableListOf<String>()
+
+  override fun visitParameter(name: String, access: Int) {
+    params.add(name)
+  }
+
+  override fun visitEnd() {
+    paramNames.add(params)
+    super.visitEnd()
+  }
+}
+
+/**
+ * Creates a list of ByteArrays of each .class file in a .jar file
+ *
+ * @param jarFile JarFile of java library wanting to be converted to klib
+ * @return List of ByteArrays of each class file in the jar
+ */
+fun loadClassDataFromJar(jarFile: JarFile): Sequence<ByteArray> {
+  return jarFile.entries().iterator().asSequence()
+    .filter { it.name.endsWith(".class") }
+    .map { jarFile.getInputStream(it).readBytes() }
+}
+
+/**
+ * Helper function to load a jar file and create and populate a J2ObjC NativeIndex and return an IndexerResult with it
+ *
+ * @param jarFiles List of Java .class files to be loaded
+ * @return A IndexerResult with a J2ObjCNativeIndex that is populated from jarFiles
+ */
+fun buildJ2ObjcNativeIndex(jarFiles: List<String>): IndexerResult {
+  val jarClassData = loadClassDataFromJar(JarFile(jarFiles[0])) // Currently only reads one jar file
```

Comment:
technically the JarFile should be closed when it's no longer needed, e.g., with the "use" function:

JarFile(jarFiles[0]).use { loadClassDataFromJar(it) }

It might also be better to read and process each class before moving on to the next one, to avoid having the entire inflated Jar in memory, effectively.

Reply:
Done.